lib/Sema/CSDiagnostics.cpp

```diff
@@ -100,6 +101,27 @@ ValueDecl *RequirementFailure::getDeclRef() const {
   return ownerType->getAnyGeneric();
 }
 
+const DeclContext *RequirementFailure::getRequirementDC() const {
+  const auto &req = getRequirement();
+  auto *DC = AffectedDecl->getDeclContext();
+
+  do {
+    auto *D = DC->getInnermostDeclarationDeclContext();
+    if (!D)
+      break;
+
+    if (auto *GC = D->getAsGenericContext()) {
+      auto *sig = GC->getGenericSignature();
```

Comment:
This looks like you just want DC->getGenericSignatureOfContext() and avoid all this?

Reply:
I don't think I can because `getGenericSignatureOfContext` is going to traverse parents until it finds one with generic signature and it would only return signature from it, but I wouldn't know which parent that was.

Reply:
getGenericSignatureOfContext() does not traverse parents, but non-generic decls in generic context inherit the generic signature of their parent, so getGenericSignature() has the same behavior.

Reply:
Good to know, I'll give it a try.